src/optimizers/quantizer.cpp

```diff
@@ -0,0 +1,187 @@
+#include <cmath>
+
+#include "optimizers/quantizer.h"
+#include "tensors/tensor_allocator.h"
+#include "tensors/tensor_operators.h"
+
+#include "functional/functional.h"
+
+namespace marian {
+
+/* simulate a fixed quantization for values in data.
+ * For example:
+ * data  = [0.96, 0.73, 0.82, 0.84, 0.42, 0.29, 0.65]
+ * res   = [1   , 0.6,  0.8 , 0.8 , 0.4,  0.2 , 0.6 ]
+ *
+ * @param data contains the original data
+ * @param res will contain the resulting quantized data. set data = quant for in-place operation
+ * @param numCenters the number of quantized centers in absolute. It should be 2^(bit-1)
+ * @param S stores the scaling factor.
+ */
+void quantizeFixed(Tensor data, Tensor res, int numCenters, float S) {
+  using namespace functional;
+  float multiplier = numCenters / S;
+
+  // clip based on the scale
+  Element(_1 = clip(_2, S), res, data);
+
+  // get the quantization center
+  Element(_1 = round(_1 * multiplier), res);
+
+  // revert back to floating point representation
+  Element(_1 /= multiplier, res);
+}
+
+/* simulate a log-based quantization for values in data. The quantized value will be in the form of
+ * S*2^q For example: 
+ * data  = [0.9, 0.7, 0.5, 0.2 , 1.1] 
+ * res   = [1,   0.5, 0.5, 0.25, 1  ]
+ *
+ * @param data contains the original data
+ * @param res will contain the resulting quantized data. set data = res for in-place operation
+ * @param size the data size
+ * @param numCenters the number of quantized centers in absolute. It should be 2^(bit-1)
+ * @param S stores the scaling factor.
+ * @param base for log quantized center. Default of 2
+ */
+void quantizeLog(Tensor data, Tensor res, int numCenters, float S, float base = 2.0f) {
+  using namespace functional;
```

Comment:
`quantizeLogarithmically()`? (This is not critical, we are getting into personal-preference territory... up to you)

Reply:
edited to logQuantization()